seb_openedx/middleware.py

```diff
@@ -38,16 +43,45 @@ def process_view(self, request, view_func, view_args, view_kwargs):
                 # Second: Granular white-listing
                 access_denied = True
 
+            masquerade, masquerade_html, user = self.handle_masquerade(request, course_key)  # pylint: disable=unused-variable
             active_comps = get_enabled_permission_classes()
             for permission in active_comps:
+                if masquerade and masquerade.role != 'staff' and permission == AlwaysAllowStaff:
+                    continue
                 if permission().check(request, course_key):
                     access_denied = False
 
-        if access_denied:
-            return self.handle_access_denied(request, view_func, view_args, view_kwargs, course_key)
+            if access_denied:
+                html = six.text_type(self.handle_access_denied(request, view_func, view_args, view_kwargs, course_key))
+                if masquerade:
+                    html += masquerade_html
+                http_response = HttpResponseForbidden(html)
+                return http_response
 
         return None
 
+    def supports_preview_menu(self, request):
+        """ if request view support preview_menu or not (not fully accurate) """
+        return bool(request.resolver_match.func.__name__ == get_courseware_module().views.index.CoursewareIndex.__name__)\
+            or inspect.getmodule(request.resolver_match.func).__name__.startswith('openedx.features.course_experience')
+
+    # pylint: disable=too-many-arguments
+    def handle_masquerade(self, request, course_key):
+        """ masquerade """
+        courseware = get_courseware_module()
+        masquerade, user = courseware.masquerade.setup_masquerade(request, course_key, request.user.is_staff)
+        if masquerade:
+            context = {
+                'course': courseware.courses.get_course(course_key),
+                'supports_preview_menu': self.supports_preview_menu(request),
+                'staff_access': request.user.is_staff,
+                'masquerade': masquerade,
+            }
+            js_reposition = render_to_string('seb-reposition-wrapper-preview-menu.html', {})
+            masquerade_html = render_to_string('preview_menu.html', context) + js_reposition
```

Comment:
Concatenating pieces of HTML is always strange when we have full scale renderers. This won't be necessary